### ISSUE TITLE
Support for multiple video codecs in transcoding profile

### ISSUE THREAD

**Issue by dmitrylyzo** [jellyfin/jellyfin#5576]
I am trying to fix https://github.com/jellyfin/jellyfin-tizen/issues/69 with https://github.com/jellyfin/jellyfin-web/pull/2519.

This video is transcoded because audio isn't supported - that's right.
```
Video #1
ID                                       : 4113 (0x1011)
Menu ID                                  : 1 (0x1)
Format                                   : HEVC
Format/Info                              : High Efficiency Video Coding
Commercial name                          : HDR10
Format profile                           : Main 10@L5.1@High
Codec ID                                 : 36
Duration                                 : 1 min 2 s
Width                                    : 3 840 pixels
Height                                   : 2 160 pixels
Display aspect ratio                     : 16:9
Frame rate                               : 23.976 (24000/1001) FPS
Color space                              : YUV
Chroma subsampling                       : 4:2:0 (Type 2)
Bit depth                                : 10 bits
Color range                              : Limited
Color primaries                          : BT.2020
Transfer characteristics                 : PQ
Matrix coefficients                      : BT.2020 non-constant
Mastering display color primaries        : Display P3
Mastering display luminance              : min: 0.0050 cd/m2, max: 4000 cd/m2
Maximum Content Light Level              : 1000 cd/m2
Maximum Frame-Average Light Level        : 60 cd/m2

Audio #1
ID                                       : 4352 (0x1100)
Menu ID                                  : 1 (0x1)
Format                                   : AC-3 MLP FBA 16-ch
Format/Info                              : Audio Coding 3 + Meridian Lossless Packing FBA with 16-channel presentation
Commercial name                          : Dolby TrueHD with Dolby Atmos
Muxing mode                              : Stream extension
Codec ID                                 : 131
Duration                                 : 1 min 2 s
Bit rate mode                            : Variable
Bit rate                                 : 640 kb/s
Maximum bit rate                         : 6 183 kb/s
Channel(s)                               : 8 channels
Channel layout                           : L R C LFE Ls Rs Lb Rb
Sampling rate                            : 48.0 kHz
Frame rate                               : 31.250 FPS (1536 SPF)
Bit depth                                : 16 bits
Compression mode                         : Lossy
Stream size                              : 4.79 MiB (1%)
Service kind                             : Complete Main
Number of dynamic objects                : 11
Bed channel count                        : 1 channel
Bed channel configuration                : LFE
```

Without my PR, transcoding profile for `ts` is:
```
TranscodingProfile {
        Container: "ts",
        Type: Video,
        VideoCodec: "h264",
        AudioCodec: "aac,mp3,ac3,eac3,opus",
        Protocol: "hls",
        EstimateContentLength: False,
        EnableMpegtsM2TsMode: False,
        TranscodeSeekInfo: Auto,
        CopyTimestamps: False,
        Context: Streaming,
        EnableSubtitlesInManifest: False,
        MaxAudioChannels: "6",
        MinSegments: 1,
        SegmentLength: 0,
        BreakOnNonKeyFrames: False
    }
```
and `hevc` is transcoded to `h264` - that's right too.

With my PR, transcoding profile for `ts` is:
```
TranscodingProfile {
        Container: "ts",
        Type: Video,
        VideoCodec: "h264,hevc",
        AudioCodec: "aac,mp3,ac3,eac3,opus",
        Protocol: "hls",
        EstimateContentLength: False,
        EnableMpegtsM2TsMode: False,
        TranscodeSeekInfo: Auto,
        CopyTimestamps: False,
        Context: Streaming,
        EnableSubtitlesInManifest: False,
        MaxAudioChannels: "6",
        MinSegments: 1,
        SegmentLength: 0,
        BreakOnNonKeyFrames: False
    }
```
but `hevc` is still transcoded to `h264` - that isn't right.
UPD: [reason](https://github.com/jellyfin/jellyfin/issues/5576#issuecomment-806178759)

If I change codec order to `hevc,h264`, I get `Direct streaming` of `hevc` :tada: , but `h264` will be transcoded to `hevc` if audio transcoding is required (on other file) :disappointed: .

This is probably due to the fact that the server gets the first codec from the list and doesn't check all codecs for minimal transcoding.

**Server System (please complete the following information):**
 - OS: Ubuntu 18.04
 - Jellyfin Version: 20210319.3-unstable
 - Hardware settings & device: -
 - Reverse proxy: -
 - Other hardware notes: -
 
**Client System (please complete the following information):**
 - Device: Samsung SmartTV UE65NU7400
 - OS: Tizen 4.0
 - Client: Tizen
 - Browser (if Web client): 
 - Client and Browser Version: bundled web 10.8.0 (my PR) + Chrome 56

Similar issue for audio https://github.com/jellyfin/jellyfin/issues/2424

**Comment by dmitrylyzo** [comment]
I found. The reason for transcoding is HDR.
https://github.com/jellyfin/jellyfin/blob/8410a9a26601d754e3c8bed36641c990f1d0e22b/Jellyfin.Api/Helpers/DynamicHlsHelper.cs#L217-L241

**Comment by dmitrylyzo** [comment]
`master.m3u8` for Samsung UE65NU7400:
```
#EXTM3U
#EXT-X-STREAM-INF:BANDWIDTH=8069160,AVERAGE-BANDWIDTH=8069160,VIDEO-RANGE=PQ,CODECS="hvc1.2.4.L153.B0,mp4a.40.2",RESOLUTION=1920x1080,FRAME-RATE=23.976
main.m3u8?DeviceId=...&MediaSourceId=cf04a07c7a906fd75fe5bfc18fd737e9&Static=false&VideoCodec=h264,hevc&AudioCodec=aac,mp3,ac3,eac3,opus&AudioStreamIndex=1&VideoBitrate=139360000&AudioBitrate=640000&PlaySessionId=3ad64c8104d0473db98a374b48b896b2&api_key=5df55e7390fd4a08a938dfdc99dcb208&SubtitleMethod=Encode&TranscodingMaxAudioChannels=6&RequireAvc=false&Tag=89f6858951c2809982200d6589404dfe&SegmentContainer=ts&MinSegments=1&BreakOnNonKeyFrames=False&h264-profile=high,main,baseline,constrainedbaseline,high10&h264-level=51&hevc-profile=main,main10&hevc-level=183&TranscodeReasons=VideoCodecNotSupported,AudioCodecNotSupported
#EXT-X-STREAM-INF:BANDWIDTH=5097496,AVERAGE-BANDWIDTH=5097496,VIDEO-RANGE=SDR,CODECS="hvc1.1.4.L150.B0,mp4a.40.2",RESOLUTION=1920x1080,FRAME-RATE=23.976
main.m3u8?DeviceId=...&MediaSourceId=cf04a07c7a906fd75fe5bfc18fd737e9&Static=false&VideoCodec=h264,hevc&AudioCodec=aac,mp3,ac3,eac3,opus&AudioStreamIndex=1&VideoBitrate=139360000&AudioBitrate=640000&PlaySessionId=3ad64c8104d0473db98a374b48b896b2&api_key=5df55e7390fd4a08a938dfdc99dcb208&SubtitleMethod=Encode&TranscodingMaxAudioChannels=6&RequireAvc=false&Tag=89f6858951c2809982200d6589404dfe&SegmentContainer=ts&MinSegments=1&BreakOnNonKeyFrames=False&h264-profile=high,main,baseline,constrainedbaseline,high10&h264-level=51&hevc-profile=main&hevc-level=183&TranscodeReasons=VideoCodecNotSupported,AudioCodecNotSupported&AllowVideoStreamCopy=false
```
TV supports HDR but uses SDR entry :confused:

**Comment by dmitrylyzo** [comment]
I performed some more tests with Samsung Browser.

1. Master playlist contains entries with same codec (H264 or HEVC or HEVC HDR10).
- TV loads the stream with the lowest bitrate.
- Then it switches to a higher bitrate.
:heavy_check_mark: PASSED

2. Master playlist contains entries with different codecs (SDR only).
- TV loads the stream with the lowest bitrate.
- When it switches to a higher bitrate, the video freezes. Only the audio is switched.
:x: FAILED

3. Master playlist contains entries with different codecs (SDR and HDR).
- TV loads the stream with the lowest bitrate.
- When it switches to a higher bitrate, the video freezes. Only the audio is switched.
:x: FAILED

In all tests the audio codec is the same.

We probably can't rely on a master playlist with different codecs. At least with a native player. 
I am trying to add a new codec condition (`IsHdr`) - almost ready-to-review. So we can block HDR content per codec and remove the block mentioned above.

**Comment by PrplHaz4** [comment]
not stale

**Comment by dmitrylyzo** [comment]
Fixed by #7325 
![Screenshot_20220328_212944](https://user-images.githubusercontent.com/56478732/160469283-3ef8c4fb-112d-4ec8-bf5d-f514886920e1.png)
4K HEVC HDR + TrueHD on Tizen 4

**Comment by dmitrylyzo** [comment]
Found why it doesn't work again.

It started working after #7325 because the streams in the playlist had the same bitrate, so the TV chose the first stream (HDR).

The reason for the same bitrate is that only HEVC was available as a supported codec. So bitrates were calculated for HEVC->HEVC for both streams.

This was fixed in #7712 (H264 returned), and since HEVC transcoding is disabled on my server, it calculates (despite the `copy` codec) a HEVC->H264 bitrate that is larger than HEVC->HEVC (SDR stream). We are back to what it was before.
